src/utils/layout/generator/GeneratorDataSources.tsx

```diff
@@ -0,0 +1,147 @@
+import { useCurrentLanguage } from 'src/features/language/LanguageProvider';
+import { useInnerLanguageWithForcedNodeSelector } from 'src/features/language/useLanguage';
+import { Validation } from 'src/features/validation/validationContext';
+import { useMultipleDelayedSelectors } from 'src/hooks/delayedSelectors';
+import { useShallowObjectMemo } from 'src/hooks/useShallowObjectMemo';
+import { useCommitWhenFinished } from 'src/utils/layout/generator/CommitQueue';
+import { Hidden, NodesInternal, useNodesWhenReadyOrInit } from 'src/utils/layout/NodesContext';
+import { useInnerDataModelBindingTranspose } from 'src/utils/layout/useDataModelBindingTranspose';
+import { useInnerNodeFormDataSelector } from 'src/utils/layout/useNodeItem';
+import { useInnerNodeTraversalSelector } from 'src/utils/layout/useNodeTraversal';
+import type { ValidationDataSources } from 'src/features/validation';
+import type { ExpressionDataSources } from 'src/utils/layout/useExpressionDataSources';
+
+const { Provider, hooks } = createHookContext({
+  useLaxInstanceDataSources: () => useLaxInstanceDataSources(),
+  useCurrentLayoutSet: () => useCurrentLayoutSet(),
+  useDefaultDataType: () => DataModels.useDefaultDataType(),
+  useReadableDataTypes: () => DataModels.useReadableDataTypes(),
+  useExternalApis: () => useExternalApis(useApplicationMetadata().externalApiIds ?? []),
+  useNodes: () => useNodesWhenReadyOrInit(),
+  useIsForcedVisibleByDevTools: () => {
+    const devToolsIsOpen = useDevToolsStore((state) => state.isOpen);
+    const devToolsHiddenComponents = useDevToolsStore((state) => state.hiddenComponents);
+    return devToolsIsOpen && devToolsHiddenComponents !== 'hide';
+  },
+
+  useGetDataElementIdForDataType: () => DataModels.useGetDataElementIdForDataType(),
+  useValidationsProcessedLast: () => Validation.useProcessedLast(),
+
+  useCommitWhenFinished: () => useCommitWhenFinished(),
+});
+
+export const GeneratorData = {
+  Provider,
+  useExpressionDataSources,
+  useValidationDataSources,
+  useDefaultDataType: hooks.useDefaultDataType,
+  useIsForcedVisibleByDevTools: hooks.useIsForcedVisibleByDevTools,
+
+  useGetDataElementIdForDataType: hooks.useGetDataElementIdForDataType,
+  useValidationsProcessedLast: hooks.useValidationsProcessedLast,
+
+  useCommitWhenFinished: hooks.useCommitWhenFinished,
+};
+
+function useExpressionDataSources(): ExpressionDataSources {
+  const [
+    formDataSelector,
+    formDataRowsSelector,
+    attachmentsSelector,
+    optionsSelector,
+    nodeDataSelector,
+    dataSelectorForTraversal,
+    isHiddenSelector,
+  ] = useMultipleDelayedSelectors(
+    FD.useDebouncedSelectorProps(),
+    FD.useDebouncedRowsSelectorProps(),
+    NodesInternal.useAttachmentsSelectorProps(),
+    NodesInternal.useNodeOptionsSelectorProps(),
+    NodesInternal.useNodeDataSelectorProps(),
+    NodesInternal.useDataSelectorForTraversalProps(),
+    Hidden.useIsHiddenSelectorProps(),
+  );
```

Comment:
Could this be used in the (non-generator) regular `useExpressionDataSources()` as well? :thinking:

Reply:
Yes it could. As long as you add the needed provider (or move the existing one). The main performance issues are inside of the node generator due to the sheer number of usages here. In other areas like page expressions and expression validation its not gonna be used thousands of times so its not gonna impact performance very much. I guess i could also optimize the existing one by doing the same thing as here (multipledelsyedselector and using the "inner"-variants)